src/App.tsx

```diff
@@ -1,26 +1,226 @@
+        );
+      })
+      .catch(() => {
+        handleError(Errors.UPDATING, setErrorNotification);
+      })
+      .finally(() => {
+        setLoadingTodoIds(currentIds =>
+          currentIds.filter(currentId => currentId !== todoId),
+        );
+        setFocusOnChange(current => !current);
+      });
+  };
+
+  const handleRenameTodo = (
+    todoId: number,
+    updateChanges: string,
+    todoTitle: string,
+  ) => {
+    if (!updateChanges.length) {
+      handleDelete(todoId);
+
+      return;
+    }
+
+    if (updateChanges === todoTitle) {
+      setEditingTodo(null);
+
+      return;
+    }
+
+    setLoadingTodoIds(currentIds => [...currentIds, todoId]);
+
+    setTodos(currentTodos =>
+      currentTodos.map(todo =>
+        todo.id === todoId ? { ...todo, title: updateChanges } : todo,
+      ),
+    );
+
+    const updates = {
+      title: updateChanges.trim(),
+    };
+
+    updateTodo(todoId, updates)
+      .then(updatedTodo => {
+        setTodos(currentTodos =>
+          currentTodos.map(todo =>
+            todo.id === updatedTodo.id ? updatedTodo : todo,
+          ),
+        );
+        setEditingTodo(null);
+      })
+      .catch(() => {
+        handleError(Errors.UPDATING, setErrorNotification);
+        setTodos(currentTodos =>
+          currentTodos === todos ? currentTodos : todos,
+        );
+      })
+      .finally(() => setLoadingTodoIds(ids => ids.filter(id => id !== todoId)));
+  };
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">
-          React Todo App - Add and Delete
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <Header
+          todos={todos}
+          toggleTodo={toggleTodo}
+          handleSubmit={handleSubmit}
+          inputChange={inputChange}
+          loadingTodoIds={loadingTodoIds}
+          focusOnChange={focusOnChange}
+          onInputChange={setInputChange}
+        />
+
+        {!!todos.length && (
+          <TodoList
+            todos={todos}
+            filter={filter}
+            onDelete={handleDelete}
+            toggleTodo={toggleTodo}
+            loadingTodoIds={loadingTodoIds}
+            renameTodo={handleRenameTodo}
+            editingTodo={editingTodo}
+            setEditingTodo={setEditingTodo}
+          />
+        )}
+
+        {tempTodo && (
+          <TodoTemp loadingTodoIds={loadingTodoIds} tempTodo={tempTodo} />
+        )}
+
+        {!!todos.length && (
+          <Footer
+            todos={todos}
+            filter={filter}
+            onDelete={handleDelete}
+            onFilterChange={setFilter}
+          />
+        )}
```

Comment:
you can move `TempTodo` to the `TodoList` and then you won't need to write `!!todos,length && ( ... )` twice